### PR TITLE
Adds logic for relative timestamp

### LINKED ISSUE #1279: Add "relative" behaviour to Timestamp container
**Is your feature request related to a problem? Please describe.**
UX have updated the designs for the Timestamp components.

Blocked on [Psammead#277](https://github.com/bbc/psammead/issues/277).

**Describe the solution you'd like**
Add the "relative" logic to the Timestamp container. This issue will not yet integrate the relative timestamp onto the page.

We suggest adding a file within the TimestampContainer that handles inputs of datetime and outputs the relative string. This PR should not handle including these relative strings in the Timestamp Container. 

**Logic:**
The text output should vary according to the following rules:

| received timestamp is... | rendered text |
|-------------|-----------|
| < 2 min ago | 1 minute ago |
| < 60 min (1 hour) ago | X minutes ago |
| < 120 min (2 hours) ago | 1 hour ago |
| > 120 min (2 hours) ago | X hours ago |

NB, beyond 10 hours we will switch to absolute timestamps. This is covered in #1280.

Potential edge case of GMT/BST change.
“X hours ago…” should always be the “real” value (if a user started a stopwatch & let it run from the moment the article was published), should not change with the +/- hour.

**Describe alternatives you've considered**

**Testing notes**
Happy for the unit tests to cover this part of the timestamp implementation. It would be good to go through the unit test logic once the dev has completed them.

Dev insight: We should have unit tests covering each of the cases in the relative table. Note: the actual timestamp component will not reflect the relative work yet. That will be carried out in a future issue. https://github.com/bbc/simorgh/issues/1280

**Additional context**

- [x] Initially labelled with ["Refinement needed"](https://github.com/bbc/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by twinlensreflex**
Resolves #1279

**Overall change:** Adds relativeTimestamp function for creating relative timestamp.

**Code changes:**

- Added unit tests for relativeTimestamp
- Added relativeTimestamp with logic as described in #1279 :

| received timestamp is... | rendered text |
|-------------|-----------|
| < 2 min ago | 1 minute ago |
| < 60 min (1 hour) ago | X minutes ago |
| < 120 min (2 hours) ago | 1 hour ago |
| > 120 min (2 hours) ago | X hours ago |

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+describe('relativeTimestamp', () => {
+  it('returns a string which ends in ago', () => {
+    const timestamp = timestampWithDiffInMins(1);
+    const result = relativeTimestamp(timestamp);
+    expect(typeof result).toEqual('string');
+    expect(result.split(' ').pop()).toEqual('ago');
+  });
+
+  relativeBehaviour(
```
Comment (dr3) [inline src/app/containers/Timestamp/relativeTimestamp.test.js]: Probably not needed, but can we have a test for like 30 seconds ago? Also one for in the future to ensure we catch the edge case

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+const timestampWithDiffInMins = mins => timeStampGenerator(mins, 60 * 1000);
+
+const relativeBehaviour = (description, input, expectedOutput) => {
+  it(description, () => {
+    const result = relativeTimestamp(input);
+    expect(result).toEqual(expectedOutput);
+  });
+};
+
+describe('relativeTimestamp', () => {
```
Comment (dr3) [inline src/app/containers/Timestamp/relativeTimestamp.test.js:27]: Probably not needed, but can we have a test for in the future to ensure we catch the edge case

File: package-lock.json [hunk 3]
```diff
@@ -10,58 +10,58 @@
       "integrity": "sha512-OfC2uemaknXr87bdLUkWog7nYuliM9Ij5HUcajsVcMCpQrcLmtxRbVFTIqmcSkSeYRBFBRxs2FiUqFJDLdiebA==",
       "dev": true,
       "requires": {
-        "@babel/highlight": "^7.0.0"
+        "@babel/highlight": "7.0.0"
```
Comment (dr3) [inline package-lock.json]: I think your npm version might be wrong, its removing carats all over the shop 🙈

**Review by dr3 — CHANGES_REQUESTED** [review]
Just a couple things :) Great work

File: src/app/containers/Timestamp/relativeTimestamp.test.js (hunk 2 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: roll safe meme
too smart 👍

Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: but actually.. what would the expected behaviour be for a case where the timestamp is in the future? 
the expected behaviour for more than 10 hours in the past is to show an absolute timestamp but we wrote this thinking that case would be caught before this function is called.
would we want a future timestamp to be caught before this is called?
@dr3

File: package-lock.json (hunk 3 above)
Comment (twinlensreflex) [reply package-lock.json]: this one is a mystery, I'm using v8.11.3. I'll copy the `package-lock.json` from master tho.

File: src/app/containers/Timestamp/relativeTimestamp.test.js (hunk 2 above)
Comment (dr3) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: Probably one for @jimjohnsonrollings, im more just concerned it doesnt break if someone has their device clock set wrong

Comment (jimjohnsonrollings) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: Why would there be a future timestamp?

Comment (dr3) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: @jimjohnsonrollings Someone might have their phone clock set wrong, such that the device thinks the ares timestamp is in the future

Comment (jimjohnsonrollings) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: How do we handle this on responsive?  I have a vague memory that we don't use the system clock - but send our server time as part of the payload to use as 'now'.

File: src/app/containers/Timestamp/relativeTimestamp.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  const currentTime = Date.now();
+  const timeDifferenceMilliseconds = currentTime - receivedTimestamp;
+  const toMinuteDivisor = 1000 * 60;
+  const timeDifferenceMinutes = parseInt(
+    timeDifferenceMilliseconds / toMinuteDivisor,
+    10,
+  );
+
+  if (timeDifferenceMinutes > 10 * 60 || timeDifferenceMilliseconds < 0) {
+    return '';
```
Comment (dr3) [inline src/app/containers/Timestamp/relativeTimestamp.js]: Any chance we can return null here? Personal preference but I think it handles nicer in react

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  );
+
+  relativeBehaviour(
+    'returns 5 hours ago',
+    timestampWithDiffInHours(5),
+    '5 hours ago',
+  );
+
+  relativeBehaviour(
+    'returns 1 minute ago for 2 milliseconds',
```
Comment (dr3) [inline src/app/containers/Timestamp/relativeTimestamp.test.js]: Bug bare but instead of milliseconds can it be seconds? Also perhaps just use the same number of seconds for `returns 1 minute ago for 2 milliseconds` and `returns an empty string for 60 milliseconds in the future`

I think '10 seconds past and future' is clearer than 2 different amount of milliseconds which are arbitrary

**Review by dr3 — CHANGES_REQUESTED** [review]

File: src/app/containers/Timestamp/relativeTimestamp.js (hunk 4 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.js]: done 😄

File: src/app/containers/Timestamp/relativeTimestamp.test.js (hunk 5 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.test.js]: and done :)

**Review by dr3 — DISMISSED** [review]

File: src/app/containers/Timestamp/relativeTimestamp.test.js (hunk 2 above)
Comment (ChrisBAshton) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: For the record, looks like we've decided to go with an empty string for the future: https://github.com/bbc/simorgh/pull/1380/commits/4ce9feee0cdb04b1b234f9e355b4cddc6829a11a

That feels odd to me - what gets announced to screen readers?

Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.test.js:27]: https://github.com/bbc/simorgh/pull/1380/commits/a61a7f3fa50d9e35f38dd5613aaca13e8d4aa146 we're returning null now,

From what I understand, instead of a relative timestamp the absolute timestamp will be displayed instead.

File: src/app/containers/Timestamp/relativeTimestamp.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+const relativeTime = receivedTimestamp => {
+  const currentTime = Date.now();
+  const timeDifferenceMilliseconds = currentTime - receivedTimestamp;
+  const toMinuteDivisor = 1000 * 60;
+  const timeDifferenceMinutes = parseInt(
+    timeDifferenceMilliseconds / toMinuteDivisor,
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.js]:
```suggestion
    timeDifferenceMilliseconds / (1000 * 60),
```

I'd avoid the one-off `const` here - its name is quite tricky to mentally parse 🤔

File: src/app/containers/Timestamp/relativeTimestamp.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
+const relativeTime = receivedTimestamp => {
+  const currentTime = Date.now();
+  const timeDifferenceMilliseconds = currentTime - receivedTimestamp;
+  const toMinuteDivisor = 1000 * 60;
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.js]:
```suggestion
```

File: src/app/containers/Timestamp/relativeTimestamp.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  const timeDifferenceMinutes = parseInt(
+    timeDifferenceMilliseconds / toMinuteDivisor,
+    10,
+  );
+
+  if (timeDifferenceMinutes > 10 * 60 || timeDifferenceMilliseconds < 0) {
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.js]:
```suggestion
  const isInFuture = timeDifferenceMilliseconds < 0;
  const isMoreThanTenHoursAgo = timeDifferenceMinutes > (10 * 60);
  if (isInFuture || isMoreThanTenHoursAgo) {
```

Here, though, would benefit from well-named one-off consts 😄

File: src/app/containers/Timestamp/relativeTimestamp.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  if (timeDifferenceMinutes >= 60) {
+    timeDiff = timeDifferenceMinutes / 60;
+    unit = timeDiff > 1 ? 'hours' : 'hour';
+  } else if (timeDifferenceMinutes <= 1) {
+    timeDiff = 1;
+    unit = 'minute';
+  } else {
+    timeDiff = timeDifferenceMinutes;
+    unit = 'minutes';
+  }
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.js:29]: I think you could avoid the middle `else if` block and just have an `else` that looks similar to your opening `if`.

```js
} else {
  timeDiff = (timeDifferenceMinutes <= 1) ? 1 : timeDifferenceMinutes;
  unit = (timeDiff === 1) ? 'minute' : 'minutes'; 
}
```

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    'returns 5 hours ago',
+    timestampWithDiffInHours(5),
+    '5 hours ago',
+  );
+
+  relativeBehaviour(
+    'returns 5 hours ago',
+    timestampWithDiffInHours(5),
+    '5 hours ago',
+  );
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.test.js:69]: This is the same as the block above?

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+  relativeBehaviour(
+    'returns 1 minute ago for 10 seconds',
+    timeStampGenerator({
+      milliseconds: 10,
+      magnitude: 1000,
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.test.js]: I think this way of generating a timestamp is fiddly to read - better to just pass what you want:

```js
timestampGenerator({ seconds: 10 });
timestampGenerator({ mins: 3 });
timestampGenerator({ hours: 3, mins: 32, seconds: 41 });
// etc
```

...rather than always passing milliseconds but sometimes passing a magnitude

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
+import relativeTimestamp from './relativeTimestamp';
+
+const timeStampGenerator = timedifference =>
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.test.js]: Pedantic of me, but it's `timestamp` rather than `timeStamp`.

**Review by ChrisBAshton — CHANGES_REQUESTED** [review]
Some suggestions to improve the code - but love the unit tests 👍

File: src/app/containers/Timestamp/relativeTimestamp.js (hunk 6 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.js]: sure, went back and forth with these because I was trying to get codeclimate to reduce the cognitive complexity, but it turns out that's just a result of having multiple operands 😭

File: src/app/containers/Timestamp/relativeTimestamp.js (hunk 9 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.js:29]: thanks 👍

File: src/app/containers/Timestamp/relativeTimestamp.test.js (hunk 11 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.test.js]: Improved this and caught an edge case 👍

File: src/app/containers/Timestamp/relativeTimestamp.js (hunk 9 above)
Comment (dr3) [reply src/app/containers/Timestamp/relativeTimestamp.js:29]: tbf it felt more readable before... 🙈

Comment (ChrisBAshton) [reply src/app/containers/Timestamp/relativeTimestamp.js:29]: Ha 😄I think we're just halfway through a refactor.

My next suggestion would be to have this on line 16:

```js
return constructTimestamp(timeDifferenceMinutes);
```

...and move lines 16-27 into that new function.

File: src/app/containers/Timestamp/relativeTimestamp.test.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    minutes: 60 * 1000,
+    seconds: 1000,
+    milliseconds: 1,
+  };
+  let timestamp = Date.now();
+  const keyNames = Object.keys(timeDifference);
+
+  keyNames.forEach(diff => {
+    timestamp -= timeDifference[diff] * magnitudes[diff];
+  });
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.test.js:15]: Nice! 👍

**Review by ChrisBAshton — DISMISSED** [review]
Happy to approve this now but also have a suggestion for reducing the cognitive complexity of `relativeTime` if we wanna make CodeClimate (and Drew!) happy.

**Comment by twinlensreflex** [comment]
@ChrisBAshton I'm open to suggestions about fixing codeclimate! I have tried _a lot_ of things and not managed to get the complexity below 5.

**Review by dr3 — APPROVED** [review]

File: src/app/containers/Timestamp/relativeTimestamp.js [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  if (timeDiffMins >= 60) {
+    timeDiff = parseInt(timeDiffMins / 60, 10);
+    unit = timeDiff > 1 ? 'hours' : 'hour';
+  } else {
+    timeDiff = timeDiffMins <= 1 ? 1 : timeDiffMins;
+    unit = timeDiff === 1 ? 'minute' : 'minutes';
+  }
+
+  return `${timeDiff} ${unit} ago`;
+};
```
Comment (ChrisBAshton) [inline src/app/containers/Timestamp/relativeTimestamp.js:14]: CodeClimate has now reduced complexity from 10 to 8, and is now explicitly complaining about the `constructTimestamp` function. So already this has helped improve things _and_ make things more specific.

If we want all the green ticks, we could work on refactoring this function.

My instinct would be this as a first step:

```js
const suffix = 'ago';
if (timeDiffMins >= 60) {
  const hours = parseInt(timeDiffMins / 60, 10);
  return `${hours} ${hours === 1 ? 'hour' : 'hours'} ${suffix}`;
}
return `${timeDiffMins} ${timeDiffMins === 1 ? 'minute' : 'minutes'} ${suffix}`;
```

...then I'd see the two return statements are extremely similar and maybe move those out to a separate function, should keep CodeClimate happy.

But I'm also happy approving this as-is, think it's readable enough 👍

**Review by ChrisBAshton — APPROVED** [review]
Approved, but with optional comment to refactor if you like!

File: src/app/containers/Timestamp/relativeTimestamp.js (hunk 14 above)
Comment (twinlensreflex) [reply src/app/containers/Timestamp/relativeTimestamp.js:14]: We need line 9
(`timeDiff = timeDiffMins <= 1 ? 1 : timeDiffMins;`)
for this test https://github.com/bbc/simorgh/blob/1a5ffed6d7beedc53bbcfcb8e44c22bb9bc40ffc/src/app/containers/Timestamp/relativeTimestamp.test.js#L71
otherwise we'll get `1 minutes ago` :(
